### PR TITLE
Promtool: Validate service discovery files

### LINKED ISSUE #8935: Promtool should validate service discovery files
## Proposal
Offer a way to validate [`json` and `yml` files loaded by `file_sd_configs`](https://github.com/prometheus/prometheus/blob/main/discovery/file/file.go#L391), either by:
- updating `promtool check config`
- adding a dedicated command `promtool check file_sd_configs`

A dedicated command would be better suited because [changing the targets list dynamically](https://prometheus.io/docs/guides/file-sd/#changing-the-targets-list-dynamically) does not require a prometheus restart/reload, meaning the files themselves are not required @ prometheus start.

## Situation
[promtool check config](https://github.com/prometheus/prometheus/blob/main/cmd/promtool/main.go#L332) validates that one or more files exist for a specific [file-sd](https://prometheus.io/docs/guides/file-sd/#installing-configuring-and-running-prometheus) configuration, but their contents are not checked.

There is no way to validate the files before publishing them.

The only way to detect the issues without checking the logs is adding an alert on `prometheus_sd_file_read_errors_total`.

## Example bad config that passes check config
`prometheus.yml`

```
scrape_configs:
  - job_name: 'node'
    file_sd_configs:
    - files: ['./service_discovery/node.yml']
```

`./service_discovery/node.yml`

```
- targets:
  - 127.0.0.1:9100
  labels: { example: nocomma between:labels }
```

## Current behaviour
Output of `promtool check config`
```
Checking prometheus.yml
  SUCCESS: 0 rule files found
```

If the file is missing, a warning is issues:
```
Checking prometheus.yml
  WARNING: file "service_discovery/node.yml" for file_sd in scrape job "node" does not exist
  SUCCESS: 0 rule files found
```

**Both scenarios currently have an exit code of 0.**

### PR DESCRIPTION
**Pull request by LeviHarrison**
This PR adds support to validate service discovery files specified in a config Promtool is checking.

Fixes #8935

### CODE REVIEW

File: cmd/promtool/main.go [hunk 1]
```diff
@@ -368,6 +375,52 @@ func checkTLSConfig(tlsConfig config_util.TLSConfig) error {
...
+		}
+	default:
+		return errors.Errorf("invalid file extension: %q", ext)
+	}
+
+	_, err = checkSDOutput(targetGroups)
+	return err
+}
+
+func checkSDOutput(targetGroups []*targetgroup.Group) ([]*targetgroup.Group, error) {
```
Comment (LeviHarrison) [inline cmd/promtool/main.go:409]: Although the processed target groups aren't currently used, I wrote it anticipating adding a standalone command to check an SD output and display the result.

https://github.com/prometheus/prometheus/issues/8935#issuecomment-860730894

File: cmd/promtool/main.go [hunk 2]
```diff
@@ -368,6 +375,52 @@ func checkTLSConfig(tlsConfig config_util.TLSConfig) error {
...
+	for i, tg := range targetGroups {
+		if tg == nil {
+			err := errors.Errorf("nil target group item found (index %v)", i)
+			return nil, err
+		}
+
+		if tg.Labels == nil {
+			tg.Labels = model.LabelSet{}
+		}
+		tg.Labels["__<filepath or url>"] = "<path or URL of your targets>"
```
Comment (roidelapluie) [inline cmd/promtool/main.go:419]: Why do we need this?

File: cmd/promtool/main.go [hunk 3]
```diff
@@ -368,6 +375,52 @@ func checkTLSConfig(tlsConfig config_util.TLSConfig) error {
...
+	}
+
+	_, err = checkSDOutput(targetGroups)
+	return err
+}
+
+func checkSDOutput(targetGroups []*targetgroup.Group) ([]*targetgroup.Group, error) {
+	for i, tg := range targetGroups {
+		if tg == nil {
+			err := errors.Errorf("nil target group item found (index %v)", i)
```
Comment (roidelapluie) [inline cmd/promtool/main.go:412]: You can return directly

**Comment by roidelapluie** [comment]
cc @dgl

File: cmd/promtool/main.go (hunk 2 above)
Comment (LeviHarrison) [reply cmd/promtool/main.go:419]: I'm trying to replicate the SD code as much as possible so we give the most accurate representation of what will be available on the server.

https://github.com/prometheus/prometheus/blob/922f9babb1a6263b821cb879af4ed416c07d9469/discovery/file/file.go#L414
https://github.com/prometheus/prometheus/blob/922f9babb1a6263b821cb879af4ed416c07d9469/discovery/http/http.go#L179

Comment (roidelapluie) [reply cmd/promtool/main.go:419]: if we do this, we would probably use the SD interfaces, instead of doing  a per-sd thing.

Comment (LeviHarrison) [reply cmd/promtool/main.go:418]: What do you mean by _per-sd thing_?

Comment (LeviHarrison) [reply cmd/promtool/main.go:418]: I'm not sure this helps but I guess I'll throw in here that in the standalone validation command, which will actually use this output, we won't know if the input was from a file or from HTTP.

`curl localhost:1000 | promtool check sd`

`cat sd.yml | promtool check sd`

Either way Promtool receives a JSON or YAML encoded text blob.

Comment (roidelapluie) [reply cmd/promtool/main.go:418]: The idea of the validation command:


promtool check sd --config.file=prometheus --job=prometheus --wait-time=5s

--wait time is used is the SD is not a "refresh SD". It is the time we wait in the discovery channel to get the targets.

The output would be:

```
{targets: [ { job_name: prometheus, targets: [before_relabeling: {labels...}, after_relabeling: {labels....}]}
```

That enables to use oauth2, tls, etc, and works with any service discovery. it also enables testing relabeling.

WDYT?

Comment (LeviHarrison) [reply cmd/promtool/main.go:418]: That's pretty cool, I really like it. 

My idea behind `curl localhost:1000 | promtool check sd` was that if you were developing a file or HTTP SD, you could quickly check your output without having to set up a config file or anything. I think that still might have a place, maybe if we don't get any flags we check if there's anything in stdin, and parse that. Maybe?

Comment (roidelapluie) [reply cmd/promtool/main.go:418]: there is more in the HTTP SD that the format. There is also the headers (content-type), the security, the return code ..

Comment (LeviHarrison) [reply cmd/promtool/main.go:418]: True true, I guess we'll keep it at `promtool check sd --config.file=prometheus --job=prometheus --wait-time=5s`. For this PR though, I'll remove the `checkSDOutput` function now that we don't need it for anything else and move it into the `checkSDFile` function.

Comment (LeviHarrison) [reply cmd/promtool/main.go:418]: We also don't need the output now.

File: cmd/promtool/main.go [hunk 4]
```diff
@@ -368,6 +375,42 @@ func checkTLSConfig(tlsConfig config_util.TLSConfig) error {
...
+		if err := yaml.UnmarshalStrict(content, &targetGroups); err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("invalid file extension: %q", ext)
+	}
+
+	for i, tg := range targetGroups {
+		if tg == nil {
+			return errors.Errorf("nil target group item found (index %v)", i)
```
Comment (roidelapluie) [inline cmd/promtool/main.go:407]:
```suggestion
			return errors.Errorf("nil target group item found (index %d)", i)
```

File: cmd/promtool/main_test.go [hunk 5]
```diff
@@ -77,3 +77,44 @@ func mockServer(code int, body string) (*httptest.Server, func() *http.Request)
...
+		},
+	}
+	for _, test := range cases {
+		t.Run(test.name, func(t *testing.T) {
+			err := checkSDFile(test.file)
+			if test.err != "" {
+				require.Equalf(t, test.err, err.Error(), "Expected error %q, got %q", test.err, err.Error())
+				return
+			}
+			require.NoErrorf(t, err, "Got error %q, expected none", err)
```
Comment (roidelapluie) [inline cmd/promtool/main_test.go:117]:
```suggestion
			require.Ok(t, err)
```

Comment (LeviHarrison) [reply cmd/promtool/main_test.go:117]: `require.Ok` doesn't exist.

**Review by roidelapluie — APPROVED** [review]
Thanks!

**Pull request merged**